pyogrio/tests/test_core.py

```diff
@@ -611,3 +612,94 @@ def test_error_handling_warning(capfd, naturalearth_lowres):
         read_info(naturalearth_lowres, INVALID="YES")
 
     assert capfd.readouterr().err == ""
+
+
+def test_vsimem_listtree(naturalearth_lowres):
+    # Prepare test data in /vsimem
+    meta, _, geometry, field_data = read(naturalearth_lowres)
+    meta["spatial_index"] = False
+    meta["geometry_type"] = "MultiPolygon"
+    test_file_path = Path("/vsimem/pyogrio_test_naturalearth_lowres.gpkg")
+    test_dir_path = Path(f"/vsimem/pyogrio_dir_test/{naturalearth_lowres.stem}.gpkg")
+
+    write(test_file_path, geometry, field_data, **meta)
+    write(test_dir_path, geometry, field_data, **meta)
+
+    # Check basic listtree
+    files = vsi_listtree("/vsimem/")
+    assert test_file_path.as_posix() in files
+    assert test_dir_path.as_posix() in files
+
+    # Check listtree with pattern
+    files = vsi_listtree("/vsimem/", pattern="pyogrio_dir_test*.gpkg")
+    assert test_file_path.as_posix() not in files
+    assert test_dir_path.as_posix() in files
+
+    files = vsi_listtree("/vsimem/", pattern="pyogrio_test*.gpkg")
+    assert test_file_path.as_posix() in files
+    assert test_dir_path.as_posix() not in files
+
+    # Clean up
+    vsi_unlink(test_file_path)
+    vsi_rmtree(test_dir_path.parent)
+
+
+@pytest.mark.parametrize(
+    "path, is_fixture, exception, error_match",
+    [
+        (
+            "naturalearth_lowres_vsimem",
+            True,
+            NotADirectoryError,
+            "Path is not a directory",
+        ),
+        ("/vsimem/non-existent", False, FileNotFoundError, "Path does not exist"),
+        ("/vsimem", False, OSError, "path to in-memory file or directory is required"),
+        ("/vsimem/", False, OSError, "path to in-memory file or directory is required"),
+    ],
+)
+def test_vsimem_rmtree_error(path, is_fixture, exception, error_match, request):
+    if is_fixture:
+        path = request.getfixturevalue(path)
+
+    with pytest.raises(exception, match=error_match):
+        vsi_rmtree(path)
```

Comment:
It might be that you already discussed this with Brendan (didn't go through all previous review comments), but personally I would find this easier to read if we just write out those three cases:

```
    with pytest.raises(FileNotFoundError, match="Path does not exist"):
        vsi_rmtree("/vsimem/non-existent")

    with pytest.raises(OSError, match="path to in-memory file or directory is required"):
        vsi_rmtree("/vsimem")

    with pytest.raises(OSError, match="path to in-memory file or directory is required"):
        vsi_rmtree("/vsimem/")
```

I would say that is not too much repetition and the little duplication is worth the readability.

Reply:
I see below it was indeed a question to parametrize. But maybe you could then at least split the test in two tests, one that uses the fixture, and one that is parametrized on the other paths/exception/message combos?

Reply:
I don't have strong opinions to parametrize if it makes things less readable.  Splitting the tests (having `is_fixture` parametrized feels a bit awkward) also is OK.

Reply:
It's a pity that using fixtures combined with `@pytest.mark.parametrize` isn't better supported. However, artificially splitting the test in two tests doesn't make it clearer for e.g. new contributers (and for me in a few months ;-)) either I think... So I'll just rollback the changes if that's OK.

Reply:
Rolling back is just fine; apologies for the poor suggestion.

Reply:
No problem... I took me a while to remember the complications of using a fixture with parametrize and I only ran into them a some weeks ago when I created the PR ;-)...